### PR TITLE
Stop breaking out of loop if a non-canonical hash is found

### PR DESCRIPTION
**Pull request by dvdplm**
Re-opens https://github.com/paritytech/parity-ethereum/pull/10566

ref. https://github.com/paritytech/parity-ethereum/issues/10085

### CODE REVIEW

File: ethcore/blockchain/src/blockchain.rs [hunk 1]
```diff
@@ -1000,11 +1003,15 @@ impl BlockChain {
...
-					.take_while(|t| t.block_number <= details.number)
-					.last()
+			match self.block_hash(details.number) {
+				Some(h) if h == hash => {
+					return self.epoch_transitions()
+						.map(|(_, t)| t)
+						.take_while(|t| t.block_number <= details.number)
+						.last()
+				},
+				Some(h) => trace!(target: "blockchain", "Found non-canonical block hash {}", h),
```
Comment (tomusdrw) [inline ethcore/blockchain/src/blockchain.rs]: Can we print `h vs hash` here?

**Comment by lazaridiscom** [comment]
Be aware that this "Stop breaking out of loop" seems to do (at least as it looks for now) more harm than good.

**Comment by tomusdrw** [comment]
@lazaridiscom can you elaborate what you mean? This fixes a lookup of canonical epoch on non-canonical chains that are longer than the current canon chain.

As you've seen when you analyzed #10085 (thanks a lot for your input!) currently we see such long non-canonical chains produced on AuRa.

However, the issue in general is that it should happen really rarely, and the fact that we can't recover from this situation is only partially related to the fix in this PR.

What happens with chains in #10085 is that we see two forks that both reach finality and the never re-organize between one and the other - that should never happen in the first place and we're working on a fix.

**Comment by lazaridiscom** [comment]
Oh, that was just general input, re my reproduction setup. I think I understand your elaborations (despite my limited domain-knowledge)

I'll continue my path tomorrow - this was too much today.

(Btw: not that I have a "bad habit" to edit the text of the issues I file. This means that there is sometimes new stuff in the issues, without pings beeing send!)

**Comment by lazaridiscom** [comment]
@all, https://github.com/paritytech/parity-ethereum/issues/10712 got a draft commit, which will ideally lead tomorrow to a complete fix. cu!

**Pull request merged**